File: web/lib/usage.ts
// Token-usage arithmetic and formatting, shared by the console's usage page.
//
// One rule runs through all of it: a missing figure is never rendered as zero.
// Usage can be absent because a daemon died mid-run, because a provider does not
// report tokens, or because the run predates the ledger — and none of those mean
// the run was free.

export interface TokenBuckets {
  input_tokens: number;
  output_tokens: number;
  cache_read_tokens: number;
  cache_write_tokens: number;
}

/**
 * The windows the console offers. All are rolling rather than calendar-aligned:
 * "the last 24 hours" is what an operator means when asking what a machine has
 * burned today, and a rolling window has no week or month boundary to get subtly
 * wrong per timezone.
 */
export const USAGE_RANGES = ["24h", "7d", "30d", "all"] as const;
export type UsageRange = (typeof USAGE_RANGES)[number];

export const USAGE_RANGE_LABELS: Record<UsageRange, string> = {
  "24h": "24h",
  "7d": "7d",
  "30d": "30d",
  all: "All",
};

export const USAGE_RANGE_TITLES: Record<UsageRange, string> = {
  "24h": "Rolling last 24 hours",
  "7d": "Rolling last 7 days",
  "30d": "Rolling last 30 days",
  all: "All time",
};

/** The dimensions a usage breakdown can be grouped by. */
export const USAGE_GROUPS = [
  "agent",
  "daemon",
  "workspace",
  "issue",
  "model",
] as const;
export type UsageGroup = (typeof USAGE_GROUPS)[number];

export const USAGE_GROUP_LABELS: Record<UsageGroup, string> = {
  agent: "Agent",
  daemon: "Runtime",
  workspace: "Workspace",
  issue: "Issue",
  model: "Model",
};

/**
 * The four buckets summed. Cache reads are a peer bucket of input, not a subset
 * of it, so this adds up without counting a cached token twice.
 */
export function totalTokens(b: TokenBuckets): number {
  return (
    b.input_tokens + b.output_tokens + b.cache_read_tokens + b.cache_write_tokens
  );
}

/**
 * Share of prompt tokens served from the cache, as a 0-1 fraction. Null when no
 * prompt tokens were sent at all — 0% would claim a cache that was never
 * exercised, which is a different (and wrong) statement.
 */
export function cacheHitRate(b: TokenBuckets): number | null {
  const prompt = b.input_tokens + b.cache_read_tokens + b.cache_write_tokens;
  if (prompt === 0) return null;
  return b.cache_read_tokens / prompt;
}

export function formatPercent(fraction: number): string {
  // Floored, not rounded: on a cache-heavy run 99.6% rounds to "100%", which
  // claims every prompt token came from cache.
  return `${Math.floor(fraction * 100)}%`;
}

/** Compact token counts — 12.3k, 4.5M, 1.2B. Exact below a thousand. */
export function formatTokens(n: number): string {
  const abs = Math.abs(n);
  if (abs < 1000) return String(n);
  const units: Array<[number, string]> = [
    [1e9, "B"],
    [1e6, "M"],
    [1e3, "k"],
  ];
  for (const [size, suffix] of units) {
    if (abs >= size) {
      const scaled = n / size;
      // One decimal place, but not a trailing ".0" — "2k" reads better than
      // "2.0k" and the precision is noise at these magnitudes anyway.
      const rounded = Math.round(scaled * 10) / 10;
      return `${Number.isInteger(rounded) ? rounded : rounded.toFixed(1)}${suffix}`;
    }
  }
  return String(n);
}

/**
 * How complete the window's data is. `reported` of `total` finished runs
 * produced token figures; the rest are unknown, not zero.
 */
export interface UsageCoverage {
  reported: number;
  total: number;
  /** Runs that finished without reporting usage. */
  missing: number;
  /** True when some finished runs told us nothing. */
  incomplete: boolean;
  /** True when nothing in the window reported usage. */
  empty: boolean;
}

export function usageCoverage(summary: {
  runs_with_usage: number;
  runs_total: number;
}): UsageCoverage {
  const missing = Math.max(0, summary.runs_total - summary.runs_with_usage);
  return {
    reported: summary.runs_with_usage,
    total: summary.runs_total,
    missing,
    incomplete: missing > 0,
    empty: summary.runs_with_usage === 0,
  };
}

/**
 * A bucket label from the server ("2026-09-15" or "2026-09-15T14:00") rendered
 * for a chart axis. Parsed as a plain string, not through Date, so a
 * local-timezone shift cannot move a bar into the neighbouring day.
 */
export function bucketLabel(bucket: string, granularity: "hour" | "day"): string {
  // Split as strings rather than through Date: parsing would re-interpret the
  // server's already-localised label and could shift a bar into the day next to
  // it. An unexpected shape falls through to the raw label rather than throwing.
  const [date, time] = bucket.split("T");
  const parts = (date ?? "").split("-");
  if (parts.length !== 3) return bucket;
  const [, month, day] = parts;
  return granularity === "hour" && time ? `${month}/${day} ${time}` : `${month}/${day}`;
}

/** Total tokens of one chart point, for scaling the bar heights. */
export function seriesMax(
  points: Array<TokenBuckets & { bucket: string }>,
): number {
  return points.reduce((max, p) => Math.max(max, totalTokens(p)), 0);
}

/**
 * The four stacked segments of one bar, as a fraction of the tallest bar in the
 * window. A bucket with nothing in it returns an empty array so the chart draws
 * a baseline tick rather than a zero-height stack.
 */
export function barSegments(
  point: TokenBuckets,
  max: number,
): Array<{ key: keyof TokenBuckets; label: string; fraction: number }> {
  if (max <= 0) return [];
  return [
    { key: "input_tokens" as const, label: "Input", fraction: point.input_tokens / max },
    { key: "output_tokens" as const, label: "Output", fraction: point.output_tokens / max },
    { key: "cache_read_tokens" as const, label: "Cache read", fraction: point.cache_read_tokens / max },
    { key: "cache_write_tokens" as const, label: "Cache write", fraction: point.cache_write_tokens / max },
  ];
}

/** The page's own timezone, sent so the server cuts day buckets where the reader is. */
export function browserTimezone(): string {
  try {
    return Intl.DateTimeFormat().resolvedOptions().timeZone || "UTC";
  } catch {
    return "UTC";
  }
}

File: pkg/types/v1/usage.go
package v1

import "time"

// UnknownModel is the usage key used when a provider reports tokens without
// naming the model. Shared so the daemon and the server cannot disagree on it.
const UnknownModel = "unknown"

// TokenBuckets are the four mutually exclusive token counts a run consumes.
// InputTokens excludes both cache buckets, so the four sum to the run's total
// without counting a cached token twice. No derived figures (totals, rates) are
// carried on the wire: they are one helper away on the client, and sending them
// would create a second place for the same number to be wrong.
type TokenBuckets struct {
	InputTokens      int64 `json:"input_tokens"`
	OutputTokens     int64 `json:"output_tokens"`
	CacheReadTokens  int64 `json:"cache_read_tokens"`
	CacheWriteTokens int64 `json:"cache_write_tokens"`
}

// UsageWindow echoes the boundaries the server actually aggregated over, so the
// client labels the range with the same numbers rather than recomputing them
// (and the effective timezone, which may differ from what was requested).
type UsageWindow struct {
	Range    string    `json:"range"`
	From     time.Time `json:"from"`
	To       time.Time `json:"to"`
	Timezone string    `json:"timezone"`
}

// UsageSummary is the headline figure for a window plus its coverage.
//
// RunsTotal counts runs that finished in the window; RunsWithUsage counts those
// that reported tokens. The gap is usage we never received — a daemon that died
// mid-task, a provider that does not report tokens, or a run predating the
// ledger. Those runs are not free, so a UI showing only token sums would
// understate spend without saying so.
type UsageSummary struct {
	TokenBuckets
	RunsWithUsage int `json:"runs_with_usage"`
	RunsTotal     int `json:"runs_total"`
}

// UsagePoint is one bucket of the trend series. Bucket is a calendar label
// ("2026-09-15", or "2026-09-15T14:00" at hour granularity) already cut in the
// window's timezone.
type UsagePoint struct {
	Bucket string `json:"bucket"`
	TokenBuckets
	RunCount int `json:"run_count"`
}

// UsageBreakdownRow is one row of a leaderboard: a dimension value and what it
// consumed. Key identifies the dimension value (an id, or a provider/model
// pair); Label is what to render.
type UsageBreakdownRow struct {
	Key           string `json:"key"`
	Label         string `json:"label"`
	AvatarURL     string `json:"avatar_url,omitempty"`
	WorkspaceSlug string `json:"workspace_slug,omitempty"`
	TokenBuckets
	RunCount int `json:"run_count"`
}

// UsageSummaryResponse is the answer to GET /api/v1/usage/summary.
type UsageSummaryResponse struct {
	Window  UsageWindow  `json:"window"`
	Summary UsageSummary `json:"summary"`
}

// UsageSeriesResponse is the answer to GET /api/v1/usage/series.
type UsageSeriesResponse struct {
	Window UsageWindow  `json:"window"`
	Bucket string       `json:"bucket"`
	Points []UsagePoint `json:"points"`
}

// UsageBreakdownResponse is the answer to GET /api/v1/usage/breakdown.
type UsageBreakdownResponse struct {
	Window  UsageWindow         `json:"window"`
	GroupBy string              `json:"group_by"`
	Rows    []UsageBreakdownRow `json:"rows"`
}
